parsers/IN_HP.py

```diff
@@ -0,0 +1,163 @@
+    'CHAMERA3': GenType.HYDRO,
+    'PARBATI3': GenType.HYDRO,
+    'NJPC': GenType.HYDRO,  # NATHPA JHAKRI HEP in type source.
+    'RAMPUR': GenType.HYDRO,
+    'KOLDAM': GenType.HYDRO,
+}
+
+
+def fetch_production(zone_key=ZONE_KEY, session=None,
+                     target_datetime: datetime.datetime = None,
+                     logger: logging.Logger = logging.getLogger(__name__)):
+    r = session or requests.session()
+    if target_datetime is None:
+        url = DATA_URL
+    else:
+        raise NotImplementedError(
+            'This parser is not yet able to parse past dates')
+    res = r.get(url)
+    assert res.status_code == 200, 'Exception when fetching production for ' \
+                                   '{}: error when calling url={}'.format(
+                                       zone_key, url)
+    soup = BeautifulSoup(res.text, 'html.parser')
+    return {
+        'zoneKey': ZONE_KEY,
+        'datetime': arrow.now(TZ).datetime,
+        'production': combine_gen(get_state_gen(soup), get_isgs_gen(soup)),
+        'source': url
+    }
+
+    # TODO: Below is used for testing, remove once parser complete.
+    # with open('/tmp/hpdata.txt', 'w') as text_file:
+    #     text_file.write(res.text)
+
+    # with open('/tmp/hpdata.txt', 'r') as text_file:
+    #     soup = BeautifulSoup(text_file, 'html.parser')
+    #     return {
+    #         'zoneKey': ZONE_KEY,
+    #         'datetime': arrow.now(TZ).datetime,
+    #         'production': combine_gen(get_state_gen(soup), get_isgs_gen(soup)),
+    #         'source': url
+    #     }
+
+
+def get_state_gen(soup):
+    """Gets the total generation by type from state powerplants (MW).
+    Data is from the table titled GENERATION OF HP(Z)"""
+    table_name = 'GENERATION OF HP(Z)'
+    gen = {GenType.HYDRO.value: 0.0, GenType.UNKNOWN.value: 0.0}
+    # Read all rows except the last (Total).
+    for row in get_table_rows(soup, 'table_5', table_name)[:-1]:
+        try:
+            cols = row.find_all('td')
+            gen_type = get_gen_type(cols[0].text)
+            gen[gen_type.value] += float(cols[1].text)
+        except (AttributeError, IndexError, ValueError):
+            raise Exception(
+                'Error importing data from row: {}'.format(row))
+    return gen
+
+
+def get_isgs_gen(soup):
+    """Gets the total generation by type from ISGS powerplants (MW).
+    ISGS means Inter-State Generating Station: one owned by multiple states.
+    Data is from the table titled (B1)ISGS(HPSLDC WEB PORTAL)"""
+    table_name = '(B1)ISGS(HPSLDC WEB PORTAL)'
+    gen = {GenType.HYDRO.value: 0.0, GenType.UNKNOWN.value: 0.0}
+    # Read all rows except the first (headers) and last two (OTHERISGS and Total).
+    for row in get_table_rows(soup, 'table_4', table_name)[1:-2]:
+        try:
+            cols = row.find_all('td')
+            if not cols[0].has_attr('class'):
+                # Ignore COMPANY column.
+                del cols[0]
```

Comment:
This won't give us an ordering problem right?

Reply:
Not quite sure what you mean, what this does is ensures we always get the entry in the second column (STATION) and not the first one (COMPANY). For some rows the first `<td>` is for COMPANY (e.g. `<td rowspan="5">NHPC</td>`), but for most it is for STATION.

Reply:
Now added some explanatory comments.

Reply:
I was thinking more about the risk of mutating an object while iterating over it but tbh it probably doesn't matter much here.